x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/CacheIteratorHelper.java

```diff
@@ -23,31 +23,33 @@
  */
 public class CacheIteratorHelper<K, V> {
     private final Cache<K, V> cache;
-    private final ReleasableLock updateLock;
-    private final ReleasableLock iteratorLock;
+    private final ReleasableLock readLock;
+    private final ReleasableLock writeLock;
 
     public CacheIteratorHelper(Cache<K, V> cache) {
         this.cache = cache;
         final ReadWriteLock lock = new ReentrantReadWriteLock();
         // the lock is used in an odd manner; when iterating over the cache we cannot have modifiers other than deletes using the
         // iterator but when not iterating we can modify the cache without external locking. When making normal modifications to the cache
-        // the read lock is obtained so that we can allow concurrent modifications; however when we need to iterate over the keys or values
+        // the read lock can obtained so that we can allow concurrent modifications; however when we need to iterate over the keys or values
         // of the cache the write lock must obtained to prevent any modifications.
-        updateLock = new ReleasableLock(lock.readLock());
-        iteratorLock = new ReleasableLock(lock.writeLock());
+        // Note - the write lock is needed for concurrent modifications across Cache#put and Cache#invalidateAll
+        // see https://github.com/elastic/elasticsearch/issues/99326 for additional information
+        readLock = new ReleasableLock(lock.readLock());
+        writeLock = new ReleasableLock(lock.writeLock());
```

Comment:
Are you sure you want to make this change? The methods were named based on how they're supposed to be used, and that nuance is lost by the name change.

The new names imply that you need to obtain the read-lock to read from the cache and the write-lock in order to modify the cache. We don't want people to have that impression: locking for basic reads and/or obtaining exclusive locks for all simply writes is going to kill performance.

If I understand the analysis in https://github.com/elastic/elasticsearch/issues/99326 correctly, the issue is that we need to take the exclusive (write) lock in order to call `invalidateAll`, I think the we should have names that reflect that behaviour rather than the misleading names of `read` and `write`

Reply:
I think we can leave the name discussion in a future PR. So I have restored the original names in this PR to focus on the actual fix. Hope this works for both of you. Thanks!

Reply:
I agree with pushing it to a separate PR.
In terms of naming, I would be happy with `acquireSharedLock` and `acquireExclusiveLock`  and use the JavaDoc to explain which cache operations need which lock (& maybe there's a better name than shared?)

I think `read` / `write` are a problem because they lead to the scenario we almost had where we held an exclusive lock for a write on the critical path. The previous names tried to avoid that problem by naming the locks after the types of operations that needed each lock, but it turns out that the set of operations that need an exclusive lock is more than just "iteration". 
If we name them shared/exclusive we force authors to think about which one they need, at the cost that it's less obvious which one you need and you might have to check the docs.

Reply:
+1 to push the naming discussion to another PR, but generally I don't like the names that mask what it actually is. I think most developers understand the semantics of a read/write lock while shared and exclusive are less well defined. 

> I think read / write are a problem because they lead to the scenario we almost had where we held an exclusive lock for a write on the critical path.

That wasn't due to the naming or misunderstanding, just a mistake while trying to fix this ASAP. I actually don't think we should expose any locks at all and just wrap the calls to the cache with the appropriate lock and rename CacheIteratorHelper to CacheHelper and javadoc to why we are wrapping. i.e. ChacheHelper#put(k,v) grabs the read lock and CacheHelper#InvalidateAll() grabs the write lock such that the calling code only ever deals with helper and the helper does the right thing.